File: ~Programming_cheat_sheets/_compiled_languages/Java/java_core/_data_structures/tree/binary_tree/Tree.java
import java.util.LinkedList;
import java.util.Queue;
import java.util.Stack;

class Node {
    int key;

    Node left, right;

    Node(int key){
        this.key = key;
    }

    @Override
    public String toString(){
        return "K: " + key;
    }
}

class Tree{
    Node root;

    void insert(int key){
        Node newNode = new Node(key);

        if (root == null){
            root = newNode;
            return;
        }

        Node currentNode = root;
        Node leaf = null;

        while(currentNode !=null){
            leaf = currentNode;

            if (key > currentNode.key){
                currentNode = currentNode.right;
            }
            else if (key < currentNode.key){
                currentNode = currentNode.left;
            }
            else {
                return; // no dups
            }
        }

        if (key > leaf.key){
            leaf.right = newNode;
        }
        else if (key < leaf.key){
            leaf.left = newNode;
        }
    }

    Node insert(Node currentNode, int key){
        Node newNode = new Node(key);

        if (root == null)
            return newNode;

        if (key > currentNode.key)
            currentNode.right = insert(currentNode.right, key);
        else if (key < currentNode.key)
            currentNode.left = insert(currentNode.left, key);
        else
            return currentNode; // no dups

        return currentNode;
    }

    Node leftRotate(Node a){
        Node b = a.right;
        Node c = b.left;

        b.left = a;
        a.right = c;

        return b;
    }

    Node rightRotate(Node a){
        Node b = a.left;
        Node c = b.right;

        b.right = a;
        a.left = c;

        return b;
    }

    Node DFS(int key){
        Stack<Node> stack = new Stack<>();

        stack.push(root);

        while(!stack.isEmpty()){
            Node node = stack.pop();

            if (node.key == key) return node;
            if (node.left != null) stack.push(node.left);
            if (node.right!= null) stack.push(node.right);
        }

        return null;
    }

    Node BFS(int key){
        Queue<Node> queue = new LinkedList<>();

        queue.add(root);

        while(!queue.isEmpty()){
            Node node = queue.remove();

            if (node.key == key) return node;
            if (node.left != null) queue.add(node.left);
            if (node.right != null) queue.add(node.right);
        }

        return null;
    }

    void inOrder(Node currentNode){
        if (currentNode == null) return;
        inOrder(currentNode.left);
        System.out.println(currentNode);
        inOrder(currentNode.right);
    }
}




class Main{
    public static void main(String[] args){
        System.out.println("aaa");

        Tree tree = new Tree();

        tree.insert(1);
        tree.insert(2);
        tree.insert(3);
        tree.insert(4);
        tree.insert(5);
        tree.insert(6);

        tree.inOrder(tree.root);
    }
}

File: ~NEEDS_REVIEW/_scripting_languages/Python/data-structures-2020/tree/tree.py
# Tree

#------------------------------ Node
class Node:
    def __init__(self, key):
        self.key = key
        self.left = None
        self.right = None

    def __str__(self):
        if  self == None:
            return
        LK = None
        RK = None
        if self.left != None:
            LK = self.left.key
        if self.right != None:
            RK = self.right.key
        return "Node(key={key}, left={LK}, right={RK})"\
            .format(key=self.key, LK=LK, RK=RK)

#------------------------------ Tree
class Tree:
    def __init__(self):
        self.root = None

    def insert(self, key):
        node = Node(key)
        if self.root == None:
            self.root = node
            return

        currentNode = self.root
        currentParent = None

        while(currentNode != None):
            currentParent = currentNode
            if key > currentNode.key:
                currentNode = currentNode.right
            elif key < currentNode.key:
                currentNode = currentNode.left
            else:
                return # no dups

        if key > currentParent.key:
            currentParent.right = node
        else:
            currentParent.left = node

    def inorder(self):
        self._inorder(self.root)

    def _inorder(self, node):
        if node == None:
            return
        self._inorder(node.left)
        print(node)
        self._inorder(node.right)

#------------------------------ Main
tree = Tree()
tree.insert(100)
tree.insert(50)
tree.insert(25)
tree.insert(75)
tree.insert(150)

tree.inorder()

""" OUTPUT
Node(key=25, left=None, right=None)
Node(key=50, left=25, right=75)
Node(key=75, left=None, right=None)
Node(key=100, left=50, right=150)
Node(key=150, left=None, right=None)
"""
